yt/frontends/halo_catalog/io.py

```diff
@@ -104,9 +79,93 @@ def _count_particles(self, data_file):
         return {"halos": nhalos}
 
     def _identify_fields(self, data_file):
-        with h5py.File(data_file.filename, mode="r") as f:
-            fields = [("halos", field) for field in f]
+        with h5py.File(data_file.filename, "r") as f:
+            fields = [
+                ("halos", field) for field in f if not isinstance(f[field], h5py.Group)
+            ]
             units = dict(
                 [(("halos", field), parse_h5_attr(f[field], "units")) for field in f]
             )
         return fields, units
+
+
+class HaloDatasetIOHandler:
+    """
+    Base class for io handlers to load halo member particles.
+    """
+
+    def _read_particle_coords(self, chunks, ptf):
+        pass
+
+    def _read_particle_fields(self, dobj, ptf):
+        # separate member particle fields from scalar fields
+        scalar_fields = defaultdict(list)
+        member_fields = defaultdict(list)
+        for ptype, field_list in sorted(ptf.items()):
+            for field in field_list:
+                if (ptype, field) in self.ds.scalar_field_list:
+                    scalar_fields[ptype].append(field)
+                else:
+                    member_fields[ptype].append(field)
+
+        all_data = self._read_scalar_fields(dobj, scalar_fields)
+        all_data.update(self._read_member_fields(dobj, member_fields))
+
+        for field, field_data in all_data.items():
+            yield field, field_data
+
+    # This will be refactored.
+    _read_particle_selection = IOHandlerGadgetFOFHaloHDF5._read_particle_selection
+
+
+class IOHandlerYTHalo(HaloDatasetIOHandler, IOHandlerYTHaloCatalog):
+    _dataset_type = "ythalo"
+
+    def _identify_fields(self, data_file):
+        with h5py.File(data_file.filename, "r") as f:
+            scalar_fields = [
+                ("halos", field) for field in f if not isinstance(f[field], h5py.Group)
+            ]
+            units = dict(
+                [(("halos", field), parse_h5_attr(f[field], "units")) for field in f]
+            )
+            if "particles" in f:
+                id_fields = [("halos", field) for field in f["particles"]]
+            else:
+                id_fields = []
+
+        return scalar_fields + id_fields, scalar_fields, id_fields, units
+
+    def _read_member_fields(self, dobj, member_fields):
+        all_data = defaultdict(lambda: np.empty(dobj.particle_number, dtype=np.float64))
+        if not member_fields:
+            return all_data
```

Comment:
```suggestion
        if not member_fields:
            return all_data
        all_data = defaultdict(lambda: np.empty(dobj.particle_number, dtype=np.float64))
```

Reply:
I'm not sure I understand this one. How can we return `all_data` before defining it?

Reply:
we can't. I reviewed this before going to bed. Clearly I was being over-confident 😅